rholang/src/test/scala/coop/rchain/rholang/interpreter/CostAccountingReducerTest.scala

```diff
@@ -0,0 +1,151 @@
+    val env               = Env.makeEnv[Par](Expr(GString("1")))
+    val originalTermCost  = Chargeable[Par].cost(varTerm)
+    val initCost          = CostAccount(1000)
+    val costAlg           = CostAccountingAlg.unsafe[Coeval](initCost)
+    val res               = Reduce.substituteAndCharge[Par, Coeval](varTerm, 0, env, costAlg).attempt.value
+    assert(res.isLeft)
+    assert(costAlg.get().value.cost === (initCost.cost - Cost(originalTermCost)))
+  }
+
+  it should "stop if OutOfPhloError is returned from RSpace" in {
+    val tuplespaceAlg = new TuplespaceAlg[Task] {
+      override def produce(
+          chan: Channel,
+          data: ListChannelWithRandom,
+          persistent: Boolean
+      ): Task[Unit] =
+        Task.raiseError(OutOfPhlogistonsError)
+      override def consume(
+          binds: Seq[(BindPattern, ChannelInstance.Quote)],
+          body: ParWithRandom,
+          persistent: Boolean
+      ): Task[Unit] = Task.raiseError(OutOfPhlogistonsError)
+    }
+
+    implicit val errorLog = new ErrorLog()
+    implicit val rand     = Blake2b512Random(128)
+    implicit val costAlg  = CostAccountingAlg.unsafe[Task](CostAccount(1000))
+    val reducer           = new DebruijnInterpreter[Task, Task.Par](tuplespaceAlg, Map.empty)
+    val send              = Send(Channel(Quote(GString("x"))), Seq(Par()))
+    val test              = reducer.inj(send).attempt.runSyncUnsafe(1.second)
+    assert(test === Left(OutOfPhlogistonsError))
+  }
+
+  it should "stop interpreter threads as soon as deploy runs out of phlo" in {
+    // we want to make two contracts run in parallel where the shorter one
+    // will use up all the phlos before 2nd will hit the RSpace. We want to test
+    // that only first program will change the Tuplespace and 2nd one fails with OOPE.
+    //
+    // @"x"!("a") | new x in { @"x"!("a" ++ "a" ++ "a" ++ … ++ "a") }
+
+    val channel      = Channel(Quote(GString("x")))
+    val depth        = 500
+    val stringConcat = Seq.fill(depth)("\"a\"").mkString("++")
+    val plainSendStorageCost = ChargingRSpace.storageCostProduce(
+      channel,
+      ListChannelWithRandom(Seq(Channel(Quote(GString("a")))))
+    )
+    val nestedSendStorageCost = ChargingRSpace.storageCostProduce(
+      channel,
+      ListChannelWithRandom(Seq(Channel(Quote(GString("a" * depth)))))
+    )
+
+    val raw =
+      s"""
+        new x in { @"x"!($stringConcat) } | @"x"!("a")
+      """
+
+    val program = Interpreter.buildNormalizedTerm(new StringReader(raw)).value
+
+    implicit val rand   = Blake2b512Random(Array.empty[Byte])
+    implicit val errLog = new ErrorLog()
+    lazy val pureRSpace = ChargingRSpaceTest.createRhoISpace()
+    lazy val (_, reducer, _) =
+      RholangAndScalaDispatcher.create[Task, Task.Par](pureRSpace, Map.empty, Map.empty)
+
+    val plainSendCost = CHANNEL_EVAL_COST + Cost(Chargeable[Channel].cost(channel)) + plainSendStorageCost + SEND_EVAL_COST
```

Comment:
for this test we should use a simpler model because the focus is not on the aggregation of the cost but on the parallel execution stuff

Reply:
The focus of this test is the aggregation of the cost in the face of parallel execution.

Reply:
Aggregation of the cost in face of parallel execution meanning that you are testing the aggregation of the costs when the | operator is present. This means that you can get rid of all the other unimportant details.

For testing cost when | is present you need two operations with different costs and expect that you charged either one cost or the other but not both. You can do that with a Rholang code like this:

```rholang
@"c" ! "x" | @"c" ! "xy"
```

That's the bare minimum rholang code needed to test the aggregation of costs for | operator. Everything else is noise.

Reply:
How do you know how much phlos you need to trigger the desired behavior?